UNAVAILABLE: io exception in a Sink which accepts pulsar messages as bytes

**Describe the bug**
When trying to send a single string message to the topic occupied in a sink that treats the message as bytes, such as `RedisSink`, it occurs the following error:
```
$ bin/pulsar-admin sink status --tenant public --namespace default --name redis-test-sink
{
  "numInstances" : 1,
  "numRunning" : 0,
  "instances" : [ {
    "instanceId" : 0,
    "status" : {
      "running" : false,
      "error" : "UNAVAILABLE: io exception",
      "numRestarts" : 0,
      "numReadFromPulsar" : 0,
      "numSystemExceptions" : 0,
      "latestSystemExceptions" : [ ],
      "numSinkExceptions" : 0,
      "latestSinkExceptions" : [ ],
      "numWrittenToSink" : 0,
      "lastReceivedTime" : 0,
      "workerId" : "c-pulsar-cluster-test-fw-localhost-8080"
    }
  } ]
}
```
The functions log shows below:
```
15:26:13.565 [public/default/redis-test-sink-0] ERROR org.apache.pulsar.functions.instance.JavaInstanceRunnable - [public/default/redis-test-sink:0] Uncaught exception in Java Instance
org.apache.pulsar.client.api.SchemaSerializationException: org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'Hello': was expecting ('true', 'false' or 'null')
 at [Source: (byte[])"Hello Pulsar"; line: 1, column: 7]
        at org.apache.pulsar.client.impl.schema.JSONSchema.decode(JSONSchema.java:77) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.client.impl.MessageImpl.getValue(MessageImpl.java:247) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.source.PulsarRecord.getValue(PulsarRecord.java:74) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.readInput(JavaInstanceRunnable.java:451) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:236) [java-instance.jar:2.4.0-SNAPSHOT]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_171]
Caused by: org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'Hello': was expecting ('true', 'false' or 'null')
 at [Source: (byte[])"Hello Pulsar"; line: 1, column: 7]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1804) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:703) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.core.json.UTF8StreamJsonParser._reportInvalidToken(UTF8StreamJsonParser.java:3532) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.core.json.UTF8StreamJsonParser._handleUnexpectedValue(UTF8StreamJsonParser.java:2627) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.core.json.UTF8StreamJsonParser._nextTokenNotInObject(UTF8StreamJsonParser.java:832) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:729) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.databind.ObjectMapper._initForReading(ObjectMapper.java:4141) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4000) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.functions.runtime.shaded.com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3091) ~[java-instance.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.client.impl.schema.JSONSchema.decode(JSONSchema.java:75) ~[java-instance.jar:2.4.0-SNAPSHOT]
        ... 5 more
15:26:13.574 [public/default/redis-test-sink-0] INFO  org.apache.pulsar.functions.instance.JavaInstanceRunnable - Closing instance
15:26:13.581 [pulsar-client-io-1-1] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [test_redis] [public/default/redis-test-sink] Closed consumer
```

--- a/pulsar-io/redis/src/main/java/org/apache/pulsar/io/redis/sink/RedisSink.java
+++ b/pulsar-io/redis/src/main/java/org/apache/pulsar/io/redis/sink/RedisSink.java
@@ -49,7 +49,7 @@ import java.util.concurrent.TimeUnit;
     configClass = RedisSinkConfig.class
 )
 @Slf4j
-public class RedisSink<T> implements Sink<T> {
+public class RedisSink<T> implements Sink<byte[]> {
 
     private RedisSinkConfig redisSinkConfig;
 
@@ -61,7 +61,7 @@ public class RedisSink<T> implements Sink<T> {
 
     private int batchSize;
 
-    private List<Record<T>> incomingList;
+    private List<Record<byte[]>> incomingList;
 
     private ScheduledExecutorService flushExecutor;
 
@@ -84,7 +84,7 @@ public class RedisSink<T> implements Sink<T> {
     }
 
     @Override
-    public void write(Record<T> record) throws Exception {
+    public void write(Record<byte[]> record) throws Exception {
         int currentSize;
         synchronized (this) {
             incomingList.add(record);
@@ -108,7 +108,7 @@ public class RedisSink<T> implements Sink<T> {
 
     private void flush() {
         final Map<byte[], byte[]> recordsToSet = new ConcurrentHashMap<>();
-        final List<Record<T>> recordsToFlush;
+        final List<Record<byte[]>> recordsToFlush;
 
         synchronized (this) {
             if (incomingList.isEmpty()) {
@@ -119,7 +119,7 @@ public class RedisSink<T> implements Sink<T> {
         }
 
         if (CollectionUtils.isNotEmpty(recordsToFlush)) {
-            for (Record<T> record: recordsToFlush) {
+            for (Record<byte[]> record: recordsToFlush) {
                 try {
                     // records with null keys or values will be ignored
                     byte[] key = toBytes("key", record.getKey().orElse(null));

--- a/pulsar-io/redis/src/main/java/org/apache/pulsar/io/redis/sink/RedisSink.java
+++ b/pulsar-io/redis/src/main/java/org/apache/pulsar/io/redis/sink/RedisSink.java
@@ -49,7 +49,7 @@ import java.util.concurrent.TimeUnit;
     configClass = RedisSinkConfig.class
 )
 @Slf4j
-public class RedisSink<T> implements Sink<byte[]> {
+public class RedisSink implements Sink<byte[]> {
 
     private RedisSinkConfig redisSinkConfig;
 
@@ -122,8 +122,8 @@ public class RedisSink<T> implements Sink<byte[]> {
             for (Record<byte[]> record: recordsToFlush) {
                 try {
                     // records with null keys or values will be ignored
-                    byte[] key = toBytes("key", record.getKey().orElse(null));
-                    byte[] value = toBytes("value", record.getValue());
+                    byte[] key = record.getKey().isPresent() ? record.getKey().get().getBytes(StandardCharsets.UTF_8) : null;
+                    byte[] value = record.getValue();
                     recordsToSet.put(key, value);
                 } catch (Exception e) {
                     record.fail();
@@ -154,20 +154,5 @@ public class RedisSink<T> implements Sink<byte[]> {
             recordsToFlush.forEach(tRecord -> tRecord.fail());
             log.error("Redis mset data interrupted exception ", e);
         }
-    }
-
-    private byte[] toBytes(String src, Object obj) {
-        final byte[] result;
-        if (obj instanceof String) {
-            String s = (String) obj;
-            result = s.getBytes(StandardCharsets.UTF_8);
-        } else if (obj instanceof byte[]) {
-            result = (byte[]) obj;
-        } else if (null == obj) {
-            result = null;
-        } else {
-            throw new IllegalArgumentException(String.format("The %s for the record must be String or Bytes.", src));
-        }
-        return result;
     }
 }

--- a/pulsar-io/redis/src/test/java/org/apache/pulsar/io/redis/sink/RedisSinkTest.java
+++ b/pulsar-io/redis/src/test/java/org/apache/pulsar/io/redis/sink/RedisSinkTest.java
@@ -18,27 +18,18 @@
  */
 package org.apache.pulsar.io.redis.sink;
 
-import io.netty.buffer.ByteBuf;
-import io.netty.buffer.Unpooled;
-import lombok.Data;
-import lombok.EqualsAndHashCode;
-import lombok.ToString;
 import lombok.extern.slf4j.Slf4j;
-import org.apache.pulsar.client.api.Message;
-import org.apache.pulsar.client.api.schema.GenericRecord;
-import org.apache.pulsar.client.impl.MessageImpl;
-import org.apache.pulsar.client.impl.schema.AutoConsumeSchema;
-import org.apache.pulsar.client.impl.schema.AvroSchema;
-import org.apache.pulsar.client.impl.schema.generic.GenericSchemaImpl;
 import org.apache.pulsar.functions.api.Record;
-import org.apache.pulsar.functions.source.PulsarRecord;
+import org.apache.pulsar.functions.instance.SinkRecord;
 import org.apache.pulsar.io.redis.EmbeddedRedisUtils;
 import org.testng.annotations.AfterMethod;
 import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.Test;
 
+import java.nio.charset.StandardCharsets;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
 
 /**
  * redis Sink test
@@ -47,17 +38,6 @@ import java.util.Map;
 public class RedisSinkTest {
 
     private EmbeddedRedisUtils embeddedRedisUtils;
-
-    /**
-     * A Simple class to test redis class
-     */
-    @Data
-    @ToString
-    @EqualsAndHashCode
-    public static class Foo {
-        private String field1;
-        private String field2;
-    }
 
     @BeforeMethod
     public void setUp() throws Exception {
@@ -83,26 +63,7 @@ public class RedisSinkTest {
         RedisSink sink = new RedisSink();
 
         // prepare a foo Record
-        Foo obj = new Foo();
-        obj.setField1("FakeFiled1");
-        obj.setField2("FakeFiled1");
-        AvroSchema<Foo> schema = AvroSchema.of(Foo.class);
-
-        byte[] bytes = schema.encode(obj);
-        ByteBuf payload = Unpooled.copiedBuffer(bytes);
-        AutoConsumeSchema autoConsumeSchema = new AutoConsumeSchema();
-        autoConsumeSchema.setSchema(GenericSchemaImpl.of(schema.getSchemaInfo()));
-
-        Message<GenericRecord> message = new MessageImpl("fake_topic_name", "77:777", configs, payload, autoConsumeSchema);
-        Record<GenericRecord> record = PulsarRecord.<GenericRecord>builder()
-            .message(message)
-            .topicName("fake_topic_name")
-            .build();
-
-        log.info("foo:{}, Message.getValue: {}, record.getValue: {}",
-            obj.toString(),
-            message.getValue().toString(),
-            record.getValue().toString());
+        Record<byte[]> record = build("fakeTopic", "fakeKey", "fakeValue");
 
         // open should success
         sink.open(configs, null);
@@ -114,5 +75,30 @@ public class RedisSinkTest {
         // sleep to wait backend flush complete
         Thread.sleep(1000);
 
+    }
+
+    private Record<byte[]> build(String topic, String key, String value) {
+        // prepare a SinkRecord
+        SinkRecord<byte[]> record = new SinkRecord<>(new Record<byte[]>() {
+            @Override
+            public Optional<String> getKey() {
+                return Optional.empty();
+            }
+
+            @Override
+            public byte[] getValue() {
+                return value.getBytes(StandardCharsets.UTF_8);
+            }
+
+            @Override
+            public Optional<String> getDestinationTopic() {
+                if (topic != null) {
+                    return Optional.of(topic);
+                } else {
+                    return Optional.empty();
+                }
+            }
+        }, value.getBytes(StandardCharsets.UTF_8));
+        return record;
     }
 }

